### PR TITLE
Ability to pass custom passes from `nvq++`

### PR DESCRIPTION
**Pull request by 1tnguyen**
### Description

Ability to pass custom pass plugins (as `.so` files) from `nvq++` to `cudaq-opt` and adding its passes to the pipeline.

Resolved https://github.com/NVIDIA/cuda-quantum/issues/969

### CODE REVIEW

File: tools/nvqpp/nvq++.in [hunk 1]
```diff
@@ -409,6 +420,18 @@ while [ $# -ne 0 ]; do
...
+		CUDAQ_OPT_PLUGIN_ARGS="${CUDAQ_OPT_PLUGIN_ARGS} --load-cudaq-plugin $2"
+		shift
+		;;
+	--opt-func-pass)
+		CUDAQ_OPT_EXTRA_PASSES=$(add_pass_to_pipeline "${CUDAQ_OPT_EXTRA_PASSES}" "func.func($2)") 
+		shift
+		;;
+	--opt-module-pass)
+		CUDAQ_OPT_EXTRA_PASSES=$(add_pass_to_pipeline "${CUDAQ_OPT_EXTRA_PASSES}" "$2") 
+		shift
```
Comment (schweitzpgi) [inline tools/nvqpp/nvq++.in:433]: I'm not sure about having two command-line options here. Ideally, we can document that a generic `--opt-pass=`_foo_ command line option requires _foo_ to be in the pass pipeline syntax which is generically `dialect.opname(passname)`. (It's possible to write a pass on the `arith.constant` ops, for example.)

File: test/NVQPP/custom_pass.cpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+// clang-format off
+// RUN: nvq++ --enable-mlir --opt-plugin %cudaq_lib_dir/CustomPassPlugin.so --opt-func-pass cudaq-custom-pass  %s -o %t && %t | FileCheck %s
```
Comment (schweitzpgi) [inline test/NVQPP/custom_pass.cpp:9]: Not this PR, but we really need to move these end-to-end execution tests.

File: tools/nvqpp/nvq++.in (hunk 1 above)
Comment (1tnguyen) [reply tools/nvqpp/nvq++.in:433]: I was a bit unsure when doing this as well but I think you're right: we better let the user specify the pass in full, e.g., `func.func(my-custom-pass)`. Will fix it accordingly.

File: tools/nvqpp/nvq++.in [hunk 3]
```diff
@@ -132,6 +132,15 @@ function show_help {
...
 
+--opt-plugin <dynamic library file>
+	Load pass plugin by specifying its dynamic library.
+
+--opt-func-pass <pass name>
+	Append a function pass to the pass pipeline.	
+
+--opt-module-pass <pass name>
+	Append a module pass to the pass pipeline.
+
```
Comment (khalatepradnya) [inline tools/nvqpp/nvq++.in:143]: Q: Can we combine these into single argument?

File: tools/nvqpp/nvq++.in (hunk 1 above)
Comment (khalatepradnya) [reply tools/nvqpp/nvq++.in:433]: ahh.. I see, I will wait for the update. thanks!

File: tools/nvqpp/nvq++.in (hunk 3 above)
Comment (1tnguyen) [reply tools/nvqpp/nvq++.in:140]: We should indeed. Fixed in https://github.com/NVIDIA/cuda-quantum/pull/1034/commits/19bec7e234c865a79a5ab3028c7abaa6fbe27267.

File: tools/nvqpp/nvq++.in (hunk 1 above)
Comment (1tnguyen) [reply tools/nvqpp/nvq++.in]: Fixed in https://github.com/NVIDIA/cuda-quantum/pull/1034/commits/19bec7e234c865a79a5ab3028c7abaa6fbe27267 :)

File: test/NVQPP/custom_pass.cpp [hunk 4]
```diff
@@ -0,0 +1,32 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+// clang-format off
+// RUN: nvq++ --enable-mlir --opt-plugin %cudaq_lib_dir/CustomPassPlugin.so --opt-pass 'func.func(cudaq-custom-pass)'  %s -o %t && %t | FileCheck %s
```
Comment (khalatepradnya) [inline test/NVQPP/custom_pass.cpp:9]: Q: Is this plugin already being built?

**Review by khalatepradnya — APPROVED** [review]

File: test/NVQPP/custom_pass.cpp (hunk 4 above)
Comment (1tnguyen) [reply test/NVQPP/custom_pass.cpp:9]: Yea, it is built as [part of our unit tests](https://github.com/NVIDIA/cuda-quantum/tree/main/unittests/plugin).

Comment (schweitzpgi) [reply test/NVQPP/custom_pass.cpp:9]: Can you double check that the cmake dependency is correct? Thanks.

**Comment by amccaskey** [comment]
I know this has been merged, but one thing I'll add here is that we need to figure out how an external user can build custom pass plugins apart from CUDA Quantum (real plugins, without requiring a fork of the main repo). I think this is probably as straightforward as ensuring that our MLIR headers are installed (which they are not now).

**Comment by 1tnguyen** [comment]
> I know this has been merged, but one thing I'll add here is that we need to figure out how an external user can build custom pass plugins apart from CUDA Quantum (real plugins, without requiring a fork of the main repo). I think this is probably as straightforward as ensuring that our MLIR headers are installed (which they are not now).

Good point! I've posted https://github.com/NVIDIA/cuda-quantum/issues/1037 to track this.

**Pull request merged**